pkg/encoder/pipe.go

```diff
@@ -0,0 +1,69 @@
+package encoder
+
+import "log"
+
+type VideoPipe struct {
+	Input  chan InFrame
+	Output chan OutFrame
+	done   chan struct{}
+
+	encoder Encoder
+
+	// frame size
+	w, h int
+}
+
+// NewVideoPipe returns new video encoder pipe.
+// By default it waits for RGBA images on the input channel,
+// converts them into YUV I420 format,
+// encodes with provided video encoder, and
+// puts the result into the output channel.
+func NewVideoPipe(enc Encoder, w, h int) *VideoPipe {
+	return &VideoPipe{
+		Output: make(chan OutFrame, 10),
+		Input:  make(chan InFrame, 2),
+		done:   make(chan struct{}),
+
+		encoder: enc,
+
+		w: w,
+		h: h,
+	}
+}
+
+// Start begins video encoding pipe.
+// Should be wrapped into a goroutine.
+func (vp *VideoPipe) Start() {
+	defer func() {
+		if r := recover(); r != nil {
+			log.Println("Warn: Recovered panic in encoding ", r)
+		}
+	}()
+
+	size := int(float32(vp.w*vp.h) * 1.5)
+	yuv := make([]byte, size, size)
+
+	for img := range vp.Input {
+		// if buffer is full skip frame
+		// why the hell we need that?
+		//if len(vp.Output) >= cap(vp.Output) {
+		//	continue
+		//}
```

Comment:
@giongto35, need help with this part of the code. Could you please explain its purpose if you remember? You had it in one of the encoders.

Reply:
Right now we have 1000/60fps * 20chanLen = 333ms window until it can be overrun. Even in the "worst" case if the framerate is constant, normally it's less than 60fps, that's a lot of time.

Reply:
you mean why I convert from RGBA to YUV? Because the encoder accept YUV, so I did that

Reply:
@giongto35, I mean:
```
if len(vp.Output) >= cap(vp.Output) {
  continue
}
```
Makes no sense.

Reply:
I'm pretty sure it never worked. I assume you wanted to skip encoding when the channel overflows. As is it makes no sense as I said on the output channel, should that be on the input channel instead?